sdk/core/azure-core/inc/azure/core/cryptography/hash.hpp

```diff
@@ -45,13 +45,6 @@ namespace Azure { namespace Core { namespace Cryptography {
      */
     virtual std::vector<uint8_t> OnFinal(const uint8_t* data, std::size_t length) = 0;
 
-  protected:
-    /**
-     * @brief Constructs a default instance of `%Hash`.
-     *
-     */
-    Hash() = default;
```

Comment:
So, removing the public ctor all together - breaking change?

Reply:
No customer could ever call it. other than from a derived class. And for the derived class, nothing changes.
But, we can't delete it, because apparently there is custom constructor declaration.

Reply:
I reverted this change from the PR. Left it as `protected` (and previous PR moved it out from `public` to `protected`, but as I explained above it does not impact customer code because you can't call a public constructor of an abstract class other than from a derived class).

Reply:
> But, we can't delete it, because apparently there is custom constructor declaration.

The change you currently have is fine, but I think I am a bit confused now. Why do we need the explicit ctor declaration?

Reply:
Apparently, `Hash(Hash const&) = delete` is in the category of custom-defined constructors: https://onlinegdb.com/oygrmmNa1

Reply:
I think we really should explicitly list all the default constructors and operators at all times, for ultimate clarity and minimum surprises.

Reply:
Even if they align with the (currently) default compiler behavior.

Reply:
Because you `=delete` one, and list of default operations changes.

Reply:
Rather than the parameterless ctor being protected and defaulted, should it be private and deleted similar to `Hash(Hash const&)`?

Reply:
We can't do that - then the class will be non-inheritable, because it does not have any other constructor, and derived classes won't be able to construct the base class.

Reply:
> I think we really should explicitly list all the default constructors and operators at all times, for ultimate clarity and minimum surprises.

I always want to do the opposite, that is, defining constructors/destructors/assignment operators only when necessary.

In the case of `Hash` base class. I can see deleting copy constructor and copy assignment operator is necessary, because the inherited class usually has some kind of handle and thus is not copiable.
Since we defined copy constructor, we also need to define default (parameterless) constructor. Customers are not supposed to call this constructor directly, so it should be `protected`.
Virtual destructor is also necessary because it's perfectly reasonable that we use a pointer to `Hash` base class to hold  an inherited class object which can be `Md5Hash` or `Crc64Hash` or maybe `Sha256Hash` if we support it in the future. The actual concrete type doesn't have to be determined at compile-time, it can be determined at run-time.

So to sum up, the current definition of `Hash` base class looks perfect to me, and we don't need to make any changes to it,